v2/bundle/x509bundle/bundle_test.go

```diff
@@ -0,0 +1,216 @@
+			require.NoError(t, err)
+
+			bundle, err := Parse("example.org", fileBytes)
+			if test.expErrContains != "" {
+				require.Error(t, err)
+				assert.Contains(t, err.Error(), test.expErrContains)
+				return
+			}
+			require.NoError(t, err)
+			assert.NotNil(t, bundle)
+			assert.Len(t, bundle.X509Roots(), test.expNumRoots)
+		})
+	}
+}
+
+func Test_X509RootCRUD(t *testing.T) {
+	// Load bundle1, which contains a single certificate
+	bundle1, err := Load("example-1.org", "testdata/cert.pem")
+	require.NoError(t, err)
+	assert.Len(t, bundle1.X509Roots(), 1)
+
+	// Load bundle2, which contains 2 certificates
+	// The first certificate is the same than the one used in bundle1
+	bundle2, err := Load("example-2.org", "testdata/certs.pem")
+	require.NoError(t, err)
+	assert.Len(t, bundle2.X509Roots(), 2)
+	assert.True(t, bundle2.HasX509Root(bundle1.X509Roots()[0]))
+
+	// Adding a new root increases the roots slice length
+	bundle1.AddX509Root(bundle2.X509Roots()[1])
+	assert.Len(t, bundle1.X509Roots(), 2)
+	assert.True(t, bundle1.HasX509Root(bundle2.X509Roots()[0]))
+	assert.True(t, bundle1.HasX509Root(bundle2.X509Roots()[1]))
+
+	// If the root already exist, it should not be added again
+	bundle1.AddX509Root(bundle2.X509Roots()[0])
+	bundle1.AddX509Root(bundle2.X509Roots()[1])
+	assert.Len(t, bundle1.X509Roots(), 2)
+	assert.True(t, bundle1.HasX509Root(bundle2.X509Roots()[0]))
+	assert.True(t, bundle1.HasX509Root(bundle2.X509Roots()[1]))
+
+	// Removing a root, decreases the root slice lenght
+	cert := bundle1.X509Roots()[0]
+	bundle1.RemoveX509Root(cert)
+	assert.Len(t, bundle1.X509Roots(), 1)
+	assert.False(t, bundle1.HasX509Root(cert))
+
+	// If the root does not exist, it should keep its size
+	bundle1.RemoveX509Root(cert)
+	assert.Len(t, bundle1.X509Roots(), 1)
+	assert.False(t, bundle1.HasX509Root(cert))
+}
+
+func Test_MarshalSucceeds(t *testing.T) {
+	// Load a bundle to marshal
+	bundle, err := Load("example.org", "testdata/certs.pem")
+	require.NoError(t, err)
+
+	// Marshal the bundle
+	pemBytes, err := bundle.Marshal()
+	require.NoError(t, err)
+	require.NotNil(t, pemBytes)
+
+	// Load original bytes for comparison
+	expBytes, err := ioutil.ReadFile("testdata/certs.pem")
+	require.NoError(t, err)
+
+	//Assert the marshalled bundle is equal to the one loaded
+	assert.Equal(t, expBytes, pemBytes)
+}
+
+func Test_GetX509BundleForTrustDomainSucceeds(t *testing.T) {
+	bundle, err := Load("example.org", "testdata/certs.pem")
+	require.NoError(t, err)
+
+	b, err := bundle.GetX509BundleForTrustDomain("example.org")
+	require.NoError(t, err)
+	require.NotNil(t, b)
+	require.Equal(t, bundle, b)
+}
+
+func Test_GetX509BundleForTrustDomainFails(t *testing.T) {
+	bundle, err := Load("example.org", "testdata/certs.pem")
+	require.NoError(t, err)
+
+	b, err := bundle.GetX509BundleForTrustDomain("another-td.org")
+	require.Error(t, err)
+	require.Nil(t, b)
+}
+
+func Test_removeElementSucceeds(t *testing.T) {
```

Comment:
Unless there is a really compelling reason, I'd rather not unit test internals, especially when they are testable via the public interface.

Reply:
Got it. This was because there was no way to test "index out of bound" scenario using the public interface. I'll remove it.